File: scratch/merge_schema.py
import re

with open("prisma/schema.prisma", "r") as f:
    content = f.read()

# The content has conflict markers
# Let's find the HEAD and main sections
match = re.search(r"<<<<<<< HEAD\n(.*?)\n=======\n(.*?)\n>>>>>>> main", content, re.DOTALL)
if not match:
    print("Conflict markers not found or already resolved!")
    exit(1)

head_sec = match.group(1)
main_sec = match.group(2)

# We want to use the main_sec as the base (since it has onDelete: Cascade),
# but we need to inject the Institution relations into User model,
# and append the Institution models to the bottom.

# 1. Find User model in main_sec and add:
#   institutionStudent       InstitutionStudent?
#   institutionTutors        InstitutionTutor[]
user_pattern = r"(model User \{.*?\n\})"
user_match = re.search(user_pattern, main_sec, re.DOTALL)
if not user_match:
    print("User model not found in main section!")
    exit(1)

user_content = user_match.group(1)
# Insert fields before the closing brace
user_lines = user_content.split("\n")
insert_idx = len(user_lines) - 2 # right before "}"
user_lines.insert(insert_idx, "  institutionStudent       InstitutionStudent?")
user_lines.insert(insert_idx + 1, "  institutionTutors        InstitutionTutor[]")
new_user_content = "\n".join(user_lines)

main_sec = main_sec.replace(user_content, new_user_content)

# 2. Extract Institution models from head_sec.
# These models are: Institution, InstitutionStaff, InstitutionStudent, InstitutionTutor
inst_models = ""
for model_name in ["Institution", "InstitutionStaff", "InstitutionStudent", "InstitutionTutor"]:
    model_pattern = rf"(model {model_name} \{{.*?\n\}})"
    model_match = re.search(model_pattern, head_sec, re.DOTALL)
    if model_match:
        inst_models += "\n\n" + model_match.group(1)
    else:
        print(f"Model {model_name} not found in HEAD section!")

# Append these models to the end of main_sec
resolved_schema = main_sec + inst_models + "\n"

# Write the resolved schema back to prisma/schema.prisma
with open("prisma/schema.prisma", "w") as f:
    f.write(resolved_schema)

print("prisma/schema.prisma conflict resolved successfully!")

File: scratch/merge_schema.js
const fs = require('fs');

const content = fs.readFileSync('prisma/schema.prisma', 'utf8');

const regex = /<<<<<<< HEAD\r?\n([\s\S]*?)\r?\n=======\r?\n([\s\S]*?)\r?\n>>>>>>> main/;
const match = content.match(regex);

if (!match) {
  console.log("Conflict markers not found!");
  process.exit(1);
}

const headSec = match[1];
const mainSec = match[2];

// We want to use mainSec as base.
// 1. Find User model in mainSec and add fields.
const userRegex = /(model User \{[\s\S]*?\n\})/;
const userMatch = mainSec.match(userRegex);

if (!userMatch) {
  console.log("User model not found in main section!");
  process.exit(1);
}

const userContent = userMatch[1];
const lines = userContent.split('\n');
const insertIdx = lines.length - 2; // right before closing brace
lines.splice(insertIdx, 0, "  institutionStudent       InstitutionStudent?", "  institutionTutors        InstitutionTutor[]");
const newUserContent = lines.join('\n');

let resolvedMain = mainSec.replace(userContent, newUserContent);

// 2. Extract Institution models from headSec
const models = ["Institution", "InstitutionStaff", "InstitutionStudent", "InstitutionTutor"];
let instModels = "";

for (const model of models) {
  const modelRegex = new RegExp(`(model ${model} \\{[\\s\\S]*?\\n\\})`);
  const modelMatch = headSec.match(modelRegex);
  if (modelMatch) {
    instModels += "\n\n" + modelMatch[1];
  } else {
    console.log(`Model ${model} not found in HEAD section!`);
  }
}

const resolvedSchema = resolvedMain + instModels + "\n";
fs.writeFileSync('prisma/schema.prisma', resolvedSchema, 'utf8');
console.log("Successfully resolved prisma/schema.prisma!");
